### ISSUE TITLE
Composer module do not work in console

### ISSUE THREAD

**Issue by bocharsky-bw** [ansible/ansible-modules-extras#2935]
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

composer
##### ANSIBLE VERSION

```
ansible 2.1.1.0
  config file =
  configured module search path = Default w/o overrides
```
##### CONFIGURATION

none
##### OS / ENVIRONMENT

OS X
##### SUMMARY

The next command do not work:

``` bash
$ ansible localhost -m composer -a "working_dir=./"
```
##### STEPS TO REPRODUCE

Run `$ ansible localhost -m composer -a "working_dir=./"` inside of directory with valid composer.json file.
##### EXPECTED RESULTS

It should install composer dependencies
##### ACTUAL RESULTS

The error output with `-vvv`:

```
Loaded callback minimal of type stdout, v2.0
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: victor
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1474021242.83-74316657830341 `" && echo ansible-tmp-1474021242.83-74316657830341="` echo $HOME/.ansible/tmp/ansible-tmp-1474021242.83-74316657830341 `" ) && sleep 0'
<127.0.0.1> PUT /var/folders/5t/qb0jlm4x7g90bp__cnr1lybh0000gn/T/tmpLAYqVM TO /Users/victor/.ansible/tmp/ansible-tmp-1474021242.83-74316657830341/composer
<127.0.0.1> EXEC /bin/sh -c 'LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /Users/victor/.ansible/tmp/ansible-tmp-1474021242.83-74316657830341/composer; rm -rf "/Users/victor/.ansible/tmp/ansible-tmp-1474021242.83-74316657830341/" > /dev/null 2>&1 && sleep 0'
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/var/folders/5t/qb0jlm4x7g90bp__cnr1lybh0000gn/T/ansible_a_ligN/ansible_module_composer.py", line 232, in <module>
    main()
  File "/var/folders/5t/qb0jlm4x7g90bp__cnr1lybh0000gn/T/ansible_a_ligN/ansible_module_composer.py", line 183, in main
    available_options = get_available_options(module=module, command=command)
  File "/var/folders/5t/qb0jlm4x7g90bp__cnr1lybh0000gn/T/ansible_a_ligN/ansible_module_composer.py", line 151, in get_available_options
    command_help_json = json.loads(out)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/__init__.py", line 338, in loads
    return _default_decoder.decode(s)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/decoder.py", line 366, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/decoder.py", line 384, in raw_decode
    raise ValueError("No JSON object could be decoded")
ValueError: No JSON object could be decoded

localhost | FAILED! => {
    "changed": false,
    "failed": true,
    "invocation": {
        "module_name": "composer"
    },
    "module_stderr": "Traceback (most recent call last):\n  File \"/var/folders/5t/qb0jlm4x7g90bp__cnr1lybh0000gn/T/ansible_a_ligN/ansible_module_composer.py\", line 232, in <module>\n    main()\n  File \"/var/folders/5t/qb0jlm4x7g90bp__cnr1lybh0000gn/T/ansible_a_ligN/ansible_module_composer.py\", line 183, in main\n    available_options = get_available_options(module=module, command=command)\n  File \"/var/folders/5t/qb0jlm4x7g90bp__cnr1lybh0000gn/T/ansible_a_ligN/ansible_module_composer.py\", line 151, in get_available_options\n    command_help_json = json.loads(out)\n  File \"/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/__init__.py\", line 338, in loads\n    return _default_decoder.decode(s)\n  File \"/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/decoder.py\", line 366, in decode\n    obj, end = self.raw_decode(s, idx=_w(s, 0).end())\n  File \"/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/decoder.py\", line 384, in raw_decode\n    raise ValueError(\"No JSON object could be decoded\")\nValueError: No JSON object could be decoded\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE",
    "parsed": false
}
```

**Comment by ansibot** [comment]
@dmtrs, @resmo, ping. This issue is waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by resmo** [comment]
Thanks @bocharsky-bw for this bug report, I  was not able to reproduce (on Debian Linux):

```
 $ ansible --version
ansible 2.1.1.0
  config file = 
  configured module search path = Default w/o overrides
```

```
 $ cat composer.json 
{
    "require": {
        "monolog/monolog": "1.0.*"
    }
}
```

running ansible

```
 $ ansible -m composer -a "working_dir=./" localhost 
 [WARNING]: Host file not found: /etc/ansible/hosts

 [WARNING]: provided hosts list is empty, only localhost is available

localhost | SUCCESS => {
    "changed": true, 
    "msg": "Loading composer repositories with package information Updating dependencies - Installing monolog/monolog (1.0.2) Loading from cache Writing lock file Generating optimized autoload files", 
    "stdout": "Loading composer repositories with package information\nUpdating dependencies\n  - Installing monolog/monolog (1.0.2)\n    Loading from cache\n\nWriting lock file\nGenerating optimized autoload files\n", 
    "stdout_lines": [
        "Loading composer repositories with package information", 
        "Updating dependencies", 
        "  - Installing monolog/monolog (1.0.2)", 
        "    Loading from cache", 
        "", 
        "Writing lock file", 
        "Generating optimized autoload files"
    ]
}
```

The only difference I see would be the OS. Is there anything else "special"?

needs_info

**Comment by bocharsky-bw** [comment]
@resmo No, there isn't. I still have this error on OS X with your `composer.json` :/

UPD: I installed Ansible with `$ brew install ansible`, probably it helps.

**Comment by bborysenko** [comment]
Same issue with OS X. It looks like we can't get help massage in json format [there](https://github.com/ansible/ansible-modules-extras/blob/devel/packaging/language/composer.py#L146):

```
$ /usr/local/bin/php /usr/local/bin/composer help install --format=json
/usr/bin/env php -d allow_url_fopen=On -d detect_unicode=Off /usr/local/Cellar/composer/1.2.1/libexec/composer.phar "$@"
$ cat /usr/local/bin/composer
#!/usr/bin/env bash
/usr/bin/env php -d allow_url_fopen=On -d detect_unicode=Off /usr/local/Cellar/composer/1.2.1/libexec/composer.phar "$@"
```

**Comment by resmo** [comment]
Could you verify this isn't a brew issue and use the official composer install method? https://getcomposer.org/doc/00-intro.md#installation-linux-unix-osx

**Comment by bborysenko** [comment]
After I just copied phar to the `/usr/local/bin/composer` everything working fine:

```
mv /usr/local/bin/composer /usr/local/bin/composer_
cp /usr/local/Cellar/composer/1.2.1/libexec/composer.phar /usr/local/bin/composer
chmod +x /usr/local/bin/composer
```

But do we really have to run composer via php? Why not just :

```
def composer_command(module, command, arguments = "", options=[]):
    composer_path = module.get_bin_path("composer", True, ["/usr/local/bin"])
    cmd           = "%s %s %s %s" % (composer_path, command, " ".join(options), arguments)
    return module.run_command(cmd)
```

**Comment by resmo** [comment]
@bocharsky-bw that is a good question. but let's discuss it in an other issue (feature request) rather than in this bug report. 

I will update the docs, that there is an issue in brew with composer causing ansible to fail. Could you open an issue on https://github.com/Homebrew/brew pointing to this issue?

**Comment by resmo** [comment]
wontfix 
The issue seems caused by homebrew

**Comment by bocharsky-bw** [comment]
Thanks, guys, original `composer.phar` works fine for me too.

Could it be a problem that brew uses symlink for `composer.phar` in `/usr/local/bin`?

``` bash
ls -la /usr/local/bin/composer
lrwxr-xr-x  1 victor  admin  37 Sep 20 12:09 /usr/local/bin/composer -> ../Cellar/composer/1.2.1/bin/composer
```

Or do they provide a broken `composer.phar`?

UPD: This problem is only with Ansible. Brew's Composer works fine for me for regular use.
UPD2: If I replace composer provided by Brew with the original composer.phar - it starts working fine with Ansible too
